core/src/main/scala/com/stripe/dagon/Dag.scala

```diff
@@ -333,58 +453,87 @@ sealed abstract class Dag[N[_]] extends Serializable { self =>
       throw new NoSuchElementException(msg)
     }
 
-  /**
-   * This is only called by ensure
-   *
-   * Note, Expr must never be a Var
-   */
-  private def addExp[T](node: N[T], exp: Expr[N, T]): (Dag[N], Id[T]) = {
-    require(!exp.isVar)
-    val nodeId = Id.next[T]()
-    (copy(id2Exp = idToExp.updated(nodeId, exp)), nodeId)
-  }
-
   /**
    * ensure the given literal node is present in the Dag
    * Note: it is important that at each moment, each node has
    * at most one id in the graph. Put another way, for all
    * Id[T] in the graph evaluate(id) is distinct.
    */
-  protected def ensure[T](node: N[T]): (Dag[N], Id[T]) =
-    find(node) match {
+  protected def ensure[T](node: N[T]): (Dag[N], Id[T]) = {
+    val lit = toLiteral(node)
+    try ensureFast(lit)
+    catch {
+      case _: Throwable => //StackOverflowError should work, but not on scala.js
+        ensureRec(lit).result
+    }
+  }
+
+  /*
+   * This does recursion on the stack, which is faster, but can overflow
+   */
+  protected def ensureFast[T](lit: Literal[N, T]): (Dag[N], Id[T]) =
```

Comment:
`ensure` basically converting `Literal` to `Expr`, right? what's do you think to remove the layer of `Expr` and merge `Literal` and `Expr` together?